---
title: "Comparing bigstatsr and bigsnpr with PLINK and EIGENSOFT"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.align = "center", out.width = "70%", 
                      fig.width = 16, fig.asp = 0.7)
options(width = 85)
```

## Prerequisites

Here, we suppose that the data have been pre-processed (quality control and imputation), and is available as a PLINK binary file.

```{r}
suppressMessages({
  # devtools::install_github("privefl/bigsnpr")
  library(bigsnpr)   
  # install.packages("glue")
  library(glue)      
  # install.packages("ggplot2")
  library(ggplot2)
})
# FastPCA: https://data.broadinstitute.org/alkesgroup/EIGENSOFT/EIG-6.1.4.tar.gz
smartpca <- "../EIG-6.1.4/bin/smartpca"
plink <- download_plink()

write.table2 <- bigsnpr:::write.table2
fread2 <- function(...) data.table::fread(data.table = FALSE, ...)

NCORES <- nb_cores()
```


## Example of GWAS

In this section, we will show how to perform a GWAS with or without our packages. For simplicity, everything will be done inside R.

### With bigstastr and bigsnpr

```{r}
celiac <- snp_attach("backingfiles/celiacQC.rds")
G <- celiac$genotypes
CHR <- celiac$map$chromosome
POS <- celiac$map$physical.pos

# Pruning
system.time(
  ind.keep <- snp_pruning(G, infos.chr = CHR, 
                          exclude = snp_indLRLDR(CHR, POS),
                          ncores = NCORES)
)
```

```{r}
# PCA
system.time(
  obj.svd <- big_randomSVD(G, snp_scaleBinom(), 
                           ind.col = ind.keep,
                           ncores = NCORES)
)

# Get population from external files
pop.files <- list.files(path = "../thesis-celiac/Dubois2010_data/",
                        pattern = "cluster_*",
                        full.names = TRUE)
pop <- snp_getSampleInfos(celiac, pop.files)[[1]]
pop.names <- c("Netherlands", "Italy", "UK1", "UK2", "Finland")
# Plot PC1 & PC2
plot(obj.svd, type = "scores", coeff = 2) +
  aes(color = pop.names[pop]) +
  labs(color = "Population", title = NULL) +
  theme(legend.position = c(0.2, 0.75)) + 
  guides(colour = guide_legend(override.aes = list(size = 4)))

ggsave("figures/celiac-pca.png", width = 1124, height = 945, scale = 1/75)
```

```{r}
# GWAS
system.time(
  gwas <- big_univLogReg(G, y01.train = celiac$fam$affection - 1,
                         covar.train = obj.svd$u, ncores = NCORES)
)

labels <- 1:22; labels[c(15, 17, 19, 21)] <- ""
snp_manhattan(snp_gc(gwas), CHR, POS, 
              coeff = 2, labels = labels) + 
  geom_hline(yintercept = -log10(5e-8), col = "red") +
  labs(title = NULL) +
  coord_cartesian(ylim = c(0, 30))

ggsave("figures/celiac-gwas-cut.png", width = 1475, height = 788, scale = 1/75)
```

### With PLINK and EIGENSOFT (and R)

```{r}
# Pruning
snp_writeBed(celiac, bedfile = "plink0.bed")
prefix_bed <- "plink0"
write.table2(LD.wiki34, "LRLDR.txt")
system.time(
  system(glue("{plink} --bfile {prefix_bed}",
              " --exclude LRLDR.txt --range",
              " --indep-pairwise 50 1 0.2",
              " --threads {NCORES}"))
)

ind.keep2 <- match(scan("plink.prune.in", what = ""), celiac$map$marker.ID)
mean(ind.keep %in% ind.keep2)

# Make a pruned dataset
write.table2(as.data.frame(celiac$map[ind.keep, "marker.ID"]), "keep.in")
system.time(
  system(glue("{plink} --bfile {prefix_bed}",
              " --extract keep.in --make-bed",
              " --threads {NCORES}"))
)

# PCA with EIGENSOFT
fastpca <- function(prefix, fast = TRUE) {
  parfile <- glue("{prefix}.pca.par")
  
  ## Renaming so that Eigensoft can recognize file type
  system(glue("cp {prefix}.bim {prefix}.pedsnp"))  
  ## Renaming so that Eigensoft can recognize file type
  system(glue("cp {prefix}.fam {prefix}.pedind"))  
  writeLines(glue(
    "genotypename: {prefix}.bed
    snpname: {prefix}.pedsnp
    indivname: {prefix}.pedind
    evecoutname: {prefix}.pca.evec
    evaloutname: {prefix}.eval
    altnormstyle: NO
    numoutevec: 10
    numoutlieriter: 0
    numoutlierevec: 10
    outliersigmathresh: 6
    qtmode: 0
    fastmode: {as.integer(fast)}"
  ), con = parfile)
  ## PCA
  system(glue("{smartpca} -p {parfile}"), ignore.stdout = TRUE)
  ## removing temp files
  system(glue("rm {prefix}.pedsnp")) 
  system(glue("rm {prefix}.pedind"))
  system(glue("rm {parfile}"))
  
  evec <- data.table::fread(glue("{prefix}.pca.evec"), 
                            data.table = FALSE, skip = 1)
  
  as.matrix(evec[, 2:11])
}

system.time(
  obj.svd2 <- fastpca("plink")
)

# FastPCA is not accurate enough (see benchmark-pca.html)
round(100 * cor(obj.svd2, obj.svd$u), 1)
```

```{r}
# GWAS
write.table2(cbind(celiac$fam[, 1:2], obj.svd$u), "covar.txt")
system.time(
  system(glue("{plink} --bfile {prefix_bed}",
              " --logistic hide-covar",
              " --covar covar.txt",
              " --threads {NCORES}"))
)

gwas2 <- fread2("plink.assoc.logistic")
all.equal(gwas2$P, predict(gwas, log10 = FALSE)) 
```

## Example of Polygenic Risk Scores (PRS)

In this section, we will show how to create a PRS with or without our packages. For simplicity, everything will be done inside R.

```{r}
# Split in training/test set (80%/20%)
ind.train <- sort(sample(nrow(G), 0.8 * nrow(G)))
ind.test <- setdiff(rows_along(G), ind.train)
```

### With bigstatsr and bigsnpr

```{r}
# GWAS on training set
y01 <- celiac$fam$affection - 1
system.time(
  gwas.train <- big_univLogReg(G, y01.train = y01[ind.train],
                               ind.train = ind.train, 
                               covar.train = obj.svd$u[ind.train, ], 
                               ncores = NCORES)
)

# Clumping
system.time(
  ind.keep <- snp_clumping(G, infos.chr = celiac$map$chromosome,
                           ind.row = ind.train,
                           S = abs(gwas.train$score),
                           size = 250, # as PLINK default
                           is.size.in.bp = TRUE,
                           infos.pos = celiac$map$physical.pos,
                           ncores = NCORES)
)

# PRS
(thrs <- c(0, exp(seq(log(0.2), log(550), length.out = 50))))
lpS <- -predict(gwas.train)
system.time(
  prs <- snp_PRS(G, betas.keep = gwas.train$estim[ind.keep],
                 ind.test = ind.test,
                 ind.keep = ind.keep,
                 lpS.keep = lpS[ind.keep], 
                 thr.list = thrs)
)

# AUC
aucs <- apply(prs, 2, AUC, target = y01[ind.test])
```

### With PLINK (and R)

```{r}
# Get a file with only training set
write.table2(cbind(celiac$fam[ind.train, 1:2], 
                   obj.svd$u[ind.train, ]), "covar_train.txt")
# GWAS on training set
system.time(
  system(glue("{plink} --bfile {prefix_bed}",
              " --keep covar_train.txt",
              " --logistic hide-covar beta",
              " --covar covar_train.txt",
              " --threads {NCORES}"))
)

gwas2 <- fread2("plink.assoc.logistic")
all.equal(gwas2$P, 10^predict(gwas.train)) 

# Clumping
system.time(
  system(glue("{plink} --bfile {prefix_bed}",
              " --keep covar_train.txt",
              " --clump plink.assoc.logistic",
              " --clump-p1 1 --clump-p2 1 --clump-r2 0.2",
              " --threads {NCORES}"))
)

clump <- fread2("plink.clumped")
ind.keep2 <- match(clump$SNP, celiac$map$marker.ID)
mean(ind.keep %in% ind.keep2)

gwas2.keep <- gwas2[ind.keep, ]
write.table2(gwas2.keep, "assoc_clumped.txt")


# PRS
write.table2(celiac$fam[ind.test, 1:2], "ind_test.txt")
write.table2(cbind(paste0("S", seq_along(thrs)), 0, 10^(-thrs)),
             "range.txt")

system.time(
  system(glue("{plink} --bfile {prefix_bed}",
              " --keep ind_test.txt",
              " --extract assoc_clumped.txt",
              " --score assoc_clumped.txt 2 4 7 sum",
              " --q-score-range range.txt assoc_clumped.txt 2 9",
              " --threads {NCORES}"))
)

scores <- matrix(NA_real_, length(ind.test), length(thrs))
for (i in seq_along(thrs)) {
  scores[, i] <- fread2(glue("plink.S{i}.profile"))$SCORESUM
}

#AUC
aucs2 <- apply(scores, 2, AUC, target = y01[ind.test])

# Number of predictor for each threshold
nb.pred <- sapply(thrs, function(thr) sum(lpS[ind.keep] > thr))
bigstatsr:::MY_THEME(qplot(nb.pred, aucs, color = "A", log = "x", 
                           cex = I(3), pch = I(1),
                           xlab = "#predictors", ylab = "AUC"), coeff = 2) +
  geom_line(show.legend = FALSE) + 
  geom_point(aes(nb.pred, aucs2, color = "B"), cex = 3, pch = 3) + 
  scale_colour_manual(
    name = NULL, 
    values = c("black", "red"),
    labels = c("PRS with bigstastr\nand bigsnpr", "PRS with plink")) + 
  guides(color = guide_legend(override.aes = list(shape = c(1, 3)))) + 
  theme(legend.position = c(0.7, 0.2)) 
```

```{r}
# Cleaning
file.remove(c(list.files(pattern = "plink*|*.txt"), "keep.in"))
```

