---
title: "Imputation"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.align = "center", out.width = "70%",
                      fig.width = 15, fig.asp = 0.7)
```

## Generate missing values

```{r, message=FALSE, warning=FALSE}
library(bigsnpr)
library(dplyr)
library(ggplot2)

NCORES <- nb_cores()
```

```{r}
popres <- snp_attach("backingfiles/popresQC.rds")

G <- popres$genotypes
n <- nrow(G)
m <- ncol(G)
```

```{r}
# Compute minor allele frequencies
maf <- snp_MAF(G)

bigstatsr:::MY_THEME(ggplot(), coeff = 2) +
  geom_histogram(aes(maf)) + 
  labs(x = "Minor Allele Frequency")

ggsave("figures/hist-maf.png", width = 1300, height = 900, scale = 1/75)
```


```{r}
# Generate number of missing values by SNP with a Beta-Binomial distribution
nbNA <- VGAM::rbetabinom.ab(m, size = n, shape1 = 0.6, shape2 = 20)

sum(nbNA) / (n * m)

bigstatsr:::MY_THEME(ggplot(), coeff = 2) +
  geom_histogram(aes(nbNA), bins = 100) + 
  labs(x = "Number of missing values")

ggsave("figures/hist-NA.png", width = 1300, height = 900, scale = 1/75)
```

```{r}
# Generate indices of missing values
indNA <- cbind(
  lapply(nbNA, function(nb) {
    `if`(nb > 0, sample(n, size = nb), NULL)
  }) %>% unlist(), 
  rep(cols_along(G), nbNA)
)

# Fill a copy of the matrix with NAs (coded as 03)
GNA <- big_copy(G)
GNA[indNA] <- as.raw(3)
```


## Imputation with BEAGLE (called via bigsnpr)

```{r}
# Write a new bedfile
popresNA <- popres
popresNA$genotypes <- GNA
bedfile <- snp_writeBed(popresNA, bedfile = tempfile(fileext = ".bed"))
```

```{r}
system.time(
  bedfile2 <- snp_beagleImpute(bedfile.in = bedfile,
                               memory.max = 30,
                               ncores = NCORES)
) # 14.6 / 14.8 / 14.8 hours
```

```{r}
# Read bedfile into R and assess the errors of imputation
popres.beagle <- snp_readBed(bedfile2, backingfile = tempfile()) %>%
  snp_attach()
G.beagle <- popres.beagle$genotypes

mean(G.beagle[indNA] != G[indNA]) # 3.1 / 3.1 / 3.1 %
table(G.beagle[indNA], G[indNA])
```


## Imputation with the new method based on XGBoost

```{r}
system.time(
  infos <- snp_fastImpute(GNA, infos.chr = popres$map$chromosome, 
                          ncores = NCORES)
) # Previous runs: 42 / 47 / 47 minutes
```

```{r, fig.asp=1}
# Errors
GNA$code256 <- bigsnpr:::CODE_IMPUTE_PRED
mean(GNA[indNA] != G[indNA]) # 4.7 / 4.8 / 4.8 %
table(GNA[indNA], G[indNA])

# Plot the estimation of the errors of imputation
data_frame(col = indNA[, 2], error = GNA[indNA] != G[indNA]) %>%
  group_by(col) %>%
  summarise(nb_err = sum(error)) %>%
  mutate(nb_err_est = (infos$pError * infos$pNA * n)[col]) %>%
  ggplot() %>%
  bigstatsr:::MY_THEME(coeff = 2) + 
  geom_abline(slope = 1, intercept = 0, col = "red") + 
  geom_point(aes(nb_err, nb_err_est), alpha = 0.4) +
  labs(x = "Number of errors", y = "Estimated number of errors")

ggsave("figures/error-impute.png", width = 1000, height = 1000, scale = 1/75)
```
