---
title: "Preprocessing"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.align = "center", out.width = "80%")
options(width = 90)
```

## Quality Control

```{r}
library(bigsnpr)

plink <- download_plink()
NCORES <- nb_cores()

popresQC.bed <- snp_plinkQC(
  prefix.in = "data/POPRES_allchr",
  plink.path = plink,
  file.type = "--file", # ped/map
  geno = 0.05,
  mind = 0.05,
  maf = 0.05,
  hwe = 1e-10,
  autosome.only = TRUE
)

popresQC2.bed <- snp_plinkIBDQC(
  bedfile.in = popresQC.bed, 
  plink.path = plink, 
  ncores = NCORES
)
```

```{r}
celiacQC.bed <- snp_plinkQC(
  prefix.in = "data/FinnuncorrNLITUK1UK3hap300",
  plink.path = plink,
  geno = 0.05,
  mind = 0.05,
  maf = 0.05,
  hwe = 1e-10,
  autosome.only = TRUE
)

celiacQC2.bed <- snp_plinkIBDQC(
  celiacQC.bed, 
  plink.path = plink,
  ncores = NCORES
)
```

## Transform PLINK files in bigSNP format

```{r}
system.time(
  popresQC.rds <- snp_readBed(popresQC2.bed, "backingfiles/popresQC")
)

system.time(
  celiacQC.rds <- snp_readBed(celiacQC2.bed, "backingfiles/celiacQC")
)
```

## Imputation of Celiac dataset

```{r}
celiac <- snp_attach(celiacQC.rds)
str(celiac, max.level = 2)
```

```{r}
G <- celiac$genotypes
big_counts(G, ind.col = 1:10)
```

```{r}
# Fast imputation
system.time(
  infos <- snp_fastImpute(G, celiac$map$chromosome, ncores = NCORES)
)
```

```{r}
plot(subset(infos, pNA > 0.001), pch = 19, cex = 0.5)
pvals <- c(0.01, 0.005, 0.002, 0.001); colvals <- 2:5
idc <- lapply(seq_along(pvals), function(i) {
  curve(pvals[i] / x, from = 0, lwd = 2, 
        col = colvals[i], add = TRUE)
})
legend("topright", legend = pvals, title = "p(NA & Error)",
       col = colvals, lty = 1, lwd = 2)
```

The above plot is useful for post-processing the imputation.
No SNP (point) is above the red line. This means that no SNP has more than 1% of error. 

```{r}
big_counts(G, ind.col = 1:10)
```

```{r}
# You need to change the code of G
G$code256 <- bigsnpr:::CODE_IMPUTE_PRED
big_counts(G, ind.col = 1:10)
```

```{r}
# To make this permanent, you need to save (modify) the file on disk
celiac$genotypes$code256 <- bigsnpr:::CODE_IMPUTE_PRED
saveRDS(celiac, "backingfiles/celiacQC.rds")
```


```{r, include=FALSE}
library(gmailr)
send_message(
  mime(To = "florian.prive.21@gmail.com",
       From = "florian.prive.21@gmail.com",
       Subject = sprintf("Knitting of %s has finished",
                         knitr::current_input()))
)
```
